commit db188c8f27d295bb03d4d2786ffaf3dd3a8b5827 (refs/changes/69/1369/7)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2017-04-28T22:16:54+00:00 (2 years, 5 months ago)
    
    NAPI-373 Generate less garbage and periodically log memory usage
    Reviewed by: Alex Wilson <alex.wilson@joyent.com>
    Approved by: Alex Wilson <alex.wilson@joyent.com>

diff --git a/lib/endpoints/index.js b/lib/endpoints/index.js
index c3cf925..d385a87 100644
--- a/lib/endpoints/index.js
+++ b/lib/endpoints/index.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright 2017, Joyent, Inc.
  */
 
 /*
@@ -24,6 +24,7 @@ var toRegister = {
     '/aggregations': require('./aggregations'),
     '/fabrics/:owner_uuid/vlans': require('./fabrics/vlans'),
     '/fabrics/:owner_uuid/vlans/networks': require('./fabrics/networks'),
+    '/manage': require('./manage'),
     '/networks': require('./networks'),
     '/networks/:network_uuid/ips': require('./networks/ips'),
     '/networks/:network_uuid/nics': require('./networks/nics'),
@@ -43,10 +44,10 @@ var toRegister = {
 /*
  * Register all endpoints with the restify server
  */
-function registerEndpoints(http, log, before) {
+function registerEndpoints(http, before, log) {
     for (var t in toRegister) {
         log.debug('Registering endpoints for "%s"', t);
-        toRegister[t].register(http, before);
+        toRegister[t].register(http, before, log);
     }
 }
 
diff --git a/lib/endpoints/manage.js b/lib/endpoints/manage.js
new file mode 100644
index 0000000..97559b0
--- /dev/null
+++ b/lib/endpoints/manage.js
@@ -0,0 +1,77 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright 2017, Joyent, Inc.
+ */
+
+/*
+ * Endpoints for managing and getting info about the running NAPI server.
+ */
+
+'use strict';
+
+var restify = require('restify');
+
+
+// --- Globals
+
+var USAGE_PERIOD = 8 * 60 * 60 * 1000; // 8 hours
+
+
+// --- Internal
+
+function doCollection() {
+    var start = process.memoryUsage();
+    start.time = Date.now();
+
+    /*
+     * A single gc() call doesn't run to completion, so we call it
+     * twice to make a larger dent in the heap.
+     */
+    global.gc();
+    global.gc();
+
+    var end = process.memoryUsage();
+    end.time = Date.now();
+
+    return { start: start, end: end };
+}
+
+
+function periodicUsageLog(log) {
+    var info = process.memoryUsage();
+    log.info({ memory: info }, 'Current memory usage');
+}
+
+
+// --- Endpoints
+
+function runGC(req, res, next) {
+    if (global.gc) {
+        res.send(200, doCollection());
+        next();
+    } else {
+        next(new restify.NotImplementedError('GC not exposed to NAPI'));
+    }
+}
+
+
+function register(http, before, log) {
+    log.info({ period: USAGE_PERIOD },
+        'Starting periodic logging of memory usage');
+    setInterval(periodicUsageLog, USAGE_PERIOD, log);
+
+    http.get({ path: '/manage/gc', name: 'RunGC' },
+        before, runGC);
+    http.head({ path: '/manage/gc', name: 'RunGC' },
+        before, runGC);
+}
+
+
+module.exports = {
+    register: register
+};
diff --git a/lib/napi.js b/lib/napi.js
index ae38974..b21f084 100644
--- a/lib/napi.js
+++ b/lib/napi.js
@@ -179,7 +179,7 @@ function NAPI(opts) {
         })(req, res, route, err);
     });
 
-    endpoints.registerEndpoints(server, this.log, before);
+    endpoints.registerEndpoints(server, before, this.log);
 }
 
 util.inherits(NAPI, EventEmitter);
diff --git a/package.json b/package.json
index cb3d8d1..4a9688d 100644
--- a/package.json
+++ b/package.json
@@ -6,12 +6,12 @@
   "private": true,
   "dependencies": {
     "assert-plus": "1.0.0",
-    "bunyan": "1.8.1",
+    "bunyan": "1.8.10",
     "clone": "1.0.2",
     "cmdln": "3.0.2",
     "dashdash": "1.7.3",
     "effluent-logger": "git+https://github.com/joshwilsdon/effluent-logger.git#d662f161a07f94045ad2afb45442931511c40e51",
-    "ip6addr": "0.2.1",
+    "ip6addr": "0.2.2",
     "jsprim": "1.2.2",
     "lomstream": "1.1.0",
     "moray": "2.0.1",
diff --git a/sbin/napid b/sbin/napid
index 7849924..67d05ce 100755
--- a/sbin/napid
+++ b/sbin/napid
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright 2015 Joyent, Inc.
+# Copyright 2017 Joyent, Inc.
 #
 
 TOP=$(unset CDPATH; cd $(dirname $0)/../; pwd)
@@ -18,10 +18,11 @@ if [[ ! -d "${NODE_INSTALL}" && -d "${TOP}/build/node" ]]; then
 fi
 
 NODE=${NODE_INSTALL}/bin/node
+NODE_FLAGS=(--expose-gc --abort-on-uncaught-exception)
 
 if [[ -z "${SMF_FMRI}" ]]; then
-        ${NODE} --abort-on-uncaught-exception ${TOP}/server.js "$@" \
+        ${NODE} ${NODE_FLAGS[@]} ${TOP}/server.js "$@" \
             | ${NODE} ${TOP}/node_modules/.bin/bunyan
 else
-        ${NODE} --abort-on-uncaught-exception ${TOP}/server.js "$@"
+        ${NODE} ${NODE_FLAGS[@]} ${TOP}/server.js "$@"
 fi
